schema/decoder.go

```diff
@@ -0,0 +1,35 @@
+package schema
+
+import "context"
+
+// HasModuleCodec is an interface that modules can implement to provide a ModuleCodec.
+// Usually these modules would also implement appmodule.AppModule, but that is not included
+// to keep this package free of any dependencies.
+type HasModuleCodec interface {
+	// ModuleCodec returns a ModuleCodec for the module.
+	ModuleCodec() (ModuleCodec, error)
+}
+
+// ModuleCodec is a struct that contains the schema and a KVDecoder for a module.
+type ModuleCodec struct {
+	// Schema is the schema for the module. It is required.
+	Schema ModuleSchema
+
+	// KVDecoder is a function that decodes a key-value pair into an ObjectUpdate.
+	// If it is nil, the module doesn't support state decoding directly.
+	KVDecoder KVDecoder
+
+	// ApplyUpdate is a function that applies an ObjectUpdate to the module's state for the given context.
+	// If it is nil, the module doesn't support applying logical updates. If this function is provided
+	// then it can be used as a genesis import path.
+	ApplyUpdate ApplyUpdate
```

Comment:
Apply update provides setting an arbitrary key to an arbitrary value?

Reply:
Basically

Reply:
So as a module developer, or a streaming developer, if I'm bootstrapping streaming then I would want to somehow wire an export of module state at height _h_ to a a sequence of `ApplyUpdate` executions?

Reply:
What do you mean by streaming developer? If you're bootstrapping a module for testing you could stream a bunch of updates via `ApplyUpdate`. But this wouldn't be for bootstrapping indexers. Is that what you mean?

Reply:
I've dropped this change from this PR to move this along quicker